autopush_rs/src/server/mod.rs

```diff
@@ -510,6 +539,76 @@ impl Drop for Server {
     }
 }
 
+enum MegaphoneState {
+    Waiting,
+    Requesting(MyFuture<MegaphoneAPIResponse>),
+}
+
+struct MegaphoneUpdater {
+    srv: Rc<Server>,
+    api_url: String,
+    state: MegaphoneState,
+    timeout: Timeout,
+    poll_interval: Duration,
+    client: reqwest::unstable::async::Client,
+}
+
+impl MegaphoneUpdater {
+    fn new(uri: &str, poll_interval: Duration, srv: &Rc<Server>) -> io::Result<MegaphoneUpdater> {
+        let client = reqwest::unstable::async::Client::builder()
+            .timeout(Duration::from_secs(1))
+            .build(&srv.handle)
+            .expect("Unable to build reqwest client".into());
+        Ok(MegaphoneUpdater {
+            srv: srv.clone(),
+            api_url: String::from(uri),
+            state: MegaphoneState::Waiting,
+            timeout: Timeout::new(Duration::from_secs(30), &srv.handle)?,
+            poll_interval,
+            client,
+        })
+    }
+}
+
+impl Future for MegaphoneUpdater {
+    type Item = ();
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<(), Error> {
+        loop {
+            let new_state = match self.state {
+                MegaphoneState::Waiting => {
+                    try_ready!(self.timeout.poll());
+                    let fut = self.client.get(&self.api_url).send()
+                        .and_then(|mut response| response.json())
+                        .map_err(|_| "Unable to query/decode the API query".into());
+                    MegaphoneState::Requesting(Box::new(fut))
+                }
+                MegaphoneState::Requesting(ref mut response) => {
+                    let at = Instant::now() + self.poll_interval;
+                    match response.poll() {
+                        Ok(Async::Ready(MegaphoneAPIResponse { broadcasts })) => {
+                            debug!("Fetched broadcasts: {:?}", broadcasts);
+                            let mut broadcaster = self.srv.broadcaster.borrow_mut();
+                            for srv in Service::from_hashmap(broadcasts) {
+                                broadcaster.add_service(srv);
+                            }
+                        }
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Err(_) => {
+                            // TODO: Flag sentry that we can't poll megaphone API
+                            debug!("Failed to get response, queue again");
```

Comment:
This may want to log the error as well just in case (same w/ the `map_err` above and dropping the error)